include/numerics/dense_matrix_base.h

```diff
@@ -139,6 +140,11 @@ class DenseMatrixBase
   add (const T2 factor,
        const DenseMatrixBase<T3> & mat);
 
+  /**
+   * Return the matrix diagonal
+   */
+  DenseVector<T> diagonal() const;
```

Comment:
If I recall correctly we can't get away with a forward declare here, since `DenseVector<T>` is returned by value, the compiler needs to know the complete type. We may not actually see any failures unless dense_matrix_base.h is included all by itself somewhere...

Reply:
Based on [this SO post](https://stackoverflow.com/a/553869/4493669) I think we should be ok here

Reply:
Huh, well the whole situation is a bit more complicated than I remembered I guess, especially considering this line

    When it comes to templates, there is no absolute rule: whether you can use an incomplete type as a template parameter is dependent on the way the type is used in the template.

I guess let's just keep it as-is until some compiler tells us otherwise...